autopilot/agent.go

```diff
@@ -498,33 +509,42 @@ func (a *Agent) controller() {
 			log.Infof("Attempting to execute channel attachment "+
 				"directives: %v", spew.Sdump(chanCandidates))
 
-			// For each recommended attachment directive, we'll
-			// launch a new goroutine to attempt to carry out the
-			// directive. If any of these succeed, then we'll
-			// receive a new state update, taking us back to the
-			// top of our controller loop.
+			// Before we proceed, we'll check to see if this attempt
+			// would take us past the total number of allowed
+			// pending opens. If so, then we'll skip this round and
+			// wait for an attempt to either fail or succeed.
 			pendingMtx.Lock()
+			if uint16(len(pendingOpens))+1 >
+				a.cfg.MaxPendingOpens {
+				pendingMtx.Unlock()
+
+				log.Debugf("Reached cap of %v pending "+
+					"channel opens, will retry "+
+					"after success/failure",
+					a.cfg.MaxPendingOpens)
+				continue
+			}
```

Comment:
Ok, the logic is equivalent to what it was when inside the loop but simpler to read.

First I thought, what will happen when there is only one slot left and `chanCandidates` contains for example 3 items. In that case, all goroutines are fired and probably 2 out of 3 will fail because there are no more slots left after the connection has been established. Would it then be better to trim `chanCandidates` to the number of available slots? Maybe not, because in that case only one candidate will be tried and if it fails the others aren't tried at all in this round. Also, depending on timing, it may be that an extra slot has opened up by the time the goroutine is evaluating the pending set. But then you could also argue to always try opening channels, even if no slots are available. This will however lead to lots of failed goroutines in the default case where the pending set didn't change between launching the goroutine and establishing the connection. Coming back to what it currently does - launch all goroutines if at least one slot is available - is probably the best option.

Maybe update comment and change condition to `uint16(len(pendingOpens)) >= a.cfg.MaxPendingOpens` (same, but may be clearer because of the omission of the `+1`).